File: Count Sub Islands.py
class Solution:
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        n = len(grid2)
        m = len(grid2[0])
        visited = set()
        islands = 0

        q = []
        for i in range(n):
            for j in range(m):
                if (i, j) in visited or grid2[i][j] == 0:
                    continue

                q = [(i, j)]
                is_island = True
                while q:
                    x, y = q.pop()
                    if (x, y) in visited:
                        continue

                    visited.add((x, y))
                    if not grid1[x][y]:
                        is_island = False

                    for (x2, y2) in [(x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1)]:
                        if x2 >= n or x2 < 0 or y2 >= m or y2 < 0 or grid2[x2][y2] == 0:
                            continue
                        q.append((x2, y2))
                if is_island:
                    islands += 1
        return islands

File: Count Sub Islands.cpp
class Solution {
public:
    int countSubIslands(vector<vector<int>>& grid1, vector<vector<int>>& grid2) {
        int n = grid2.size(), m = grid2[0].size();
        set<pair<int, int>> visited;
        int islands = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                auto p = make_pair(i, j);
                if (visited.count(p) || grid2[i][j] == 0) continue;

                vector<pair<int, int>> q = {{i, j}};
                bool is_island = true;
                while (!q.empty()) {
                    auto [x, y] = q.back();
                    q.pop_back();
                    auto p2 = make_pair(x, y);
                    if (visited.contains(p2)) continue;
                    visited.insert(p2);

                    if (grid1[x][y] == 0) is_island = false;
                    vector<pair<int, int>> dirs = {
                        {x-1, y},
                        {x+1, y},
                        {x, y-1},
                        {x, y+1}
                    };

                    for (auto [x2, y2]: dirs) {
                        if (x2 >= n || x2 < 0 || y2 >= m || y2 < 0 || grid2[x2][y2] == 0) continue;
                        q.emplace_back(x2, y2);
                    }
                }
                if (is_island) islands++;
            }
        }
        return islands;
    }
};
